lightning/src/ln/channelmanager.rs

```diff
@@ -1398,121 +1408,128 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+			}
 
-				// Note that we could obviously respond immediately with an update_fulfill_htlc
-				// message, however that would leak that we are the recipient of this payment, so
-				// instead we stay symmetric with the forwarding case, only responding (after a
-				// delay) once they've send us a commitment_signed!
+			// OUR PAYMENT!
+			// final_expiry_too_soon
+			// We have to have some headroom to broadcast on chain if we have the preimage, so make sure we have at least
+			// HTLC_FAIL_BACK_BUFFER blocks to go.
+			// Also, ensure that, in the case of an unknown payment hash, our payment logic has enough time to fail the HTLC backward
+			// before our onchain logic triggers a channel closure (see HTLC_FAIL_BACK_BUFFER rational).
+			if (msg.cltv_expiry as u64) <= self.best_block.read().unwrap().height() as u64 + HTLC_FAIL_BACK_BUFFER as u64 + 1 {
+				return_err!("The final CLTV expiry is too soon to handle", 17, &[0;0]);
+			}
+			// final_incorrect_htlc_amount
+			if next_hop_data.amt_to_forward > msg.amount_msat {
+				return_err!("Upstream node sent less than we were supposed to receive in payment", 19, &byte_utils::be64_to_array(msg.amount_msat));
+			}
+			// final_incorrect_cltv_expiry
+			if next_hop_data.outgoing_cltv_value != msg.cltv_expiry {
+				return_err!("Upstream node set CLTV to the wrong value", 18, &byte_utils::be32_to_array(msg.cltv_expiry));
+			}
 
-				PendingHTLCStatus::Forward(PendingHTLCInfo {
-					routing: PendingHTLCRouting::Receive {
-						payment_data: payment_data.unwrap(),
-						incoming_cltv_expiry: msg.cltv_expiry,
-					},
-					payment_hash: msg.payment_hash.clone(),
-					incoming_shared_secret: shared_secret,
-					amt_to_forward: next_hop_data.amt_to_forward,
-					outgoing_cltv_value: next_hop_data.outgoing_cltv_value,
-				})
-			} else {
-				let mut new_packet_data = [0; 20*65];
-				let read_pos = chacha_stream.read(&mut new_packet_data).unwrap();
-				#[cfg(debug_assertions)]
-				{
-					// Check two things:
-					// a) that the behavior of our stream here will return Ok(0) even if the TLV
-					//    read above emptied out our buffer and the unwrap() wont needlessly panic
-					// b) that we didn't somehow magically end up with extra data.
-					let mut t = [0; 1];
-					debug_assert!(chacha_stream.read(&mut t).unwrap() == 0);
-				}
-				// Once we've emptied the set of bytes our peer gave us, encrypt 0 bytes until we
-				// fill the onion hop data we'll forward to our next-hop peer.
-				chacha_stream.chacha.process_in_place(&mut new_packet_data[read_pos..]);
+			let routing = match next_hop_data.format {
+				msgs::OnionHopDataFormat::Legacy { .. } => return_err!("We require payment_secrets", 0x4000|0x2000|3, &[0;0]),
+				msgs::OnionHopDataFormat::NonFinalNode { .. } => return_err!("Got non final data with an HMAC of 0", 0x4000 | 22, &[0;0]),
+				msgs::OnionHopDataFormat::FinalNode { payment_data, keysend_preimage } => {
+					if let Some(data) = payment_data {
```

Comment:
Can we add a `payment_data.is_some() && keysend_preimage.is_some()` -> "We don't support MPP keysend payments" check?